edx/edx-ora2
# [BD-05] [TNL-7310] [BB-3388] Pluggable ORA Submission Editor - Base

This PR adds pluggable submission editor support for ORA.

In order to add multiple submission editor along with pluggability following flow has been implemented -

1. ORA gets available editor configurations from ``editor_config.py`` where it's defined as -
```python
AVAILABLE_EDITORS = {
    'text': {  # key is the identification for the editor
        'display': 'Simple Text Editor',  # displays name in studio settings selectbox
        'js': ['/static/js/openassessment-editor-textarea.js'],  # this JS file tells ORA how this editor behaves 
        'css': []  # this CSS files will be included
    },
}
```
2. This ``AVAILABLE_EDITORS`` can be overridden by the ``ORA_AVAILABLE_EDITORS`` settings.
3. ORA uses the ``ResponseEditorLoader`` to load the selected editor. Selected editor and Available editor configuration are passed via JS context.
4. ``ResponseEditorLoader`` uses ``RequireJS`` to load JS files from the editor config.
4. The file that tells how a ``Editor`` behaves should have the following statucture -
```javascript
(function (define) {
  define(['additional js for this editor to work'], () => {
    class Editor {
      load(elements) {} // loads the editor

      setOnChangeListener(handler) {} // binds handler to editor's on change
      
      response(texts) {} // sets editor value if texts are present else returns existing values
    }  

    // return a function, to be able to create a new instance every time. 
    // otherwise, it can collide with other ORA block in same subsection.
    return function () {
      return new Editor();
    };
  });

// Studio has define directly on the window, LMS has RequireJS.
}).call(window, window.define || window.RequireJS.define);

```

**JIRA tickets**: 
- https://openedx.atlassian.net/browse/TNL-7310
- https://tasks.opencraft.com/browse/BB-3388

~~**Discussions**:~~

**Dependencies**: None

**Screenshots**: 
![pluggable-ora-submission-editor](https://user-images.githubusercontent.com/1010244/105444474-19a59b00-5c98-11eb-84db-5d4d9a296f3f.png)


**Sandbox URL**: https://bb3388.sandbox.opencraft.hosting/

**Merge deadline**: None

**Testing instructions**:

1. Clone this PR on your local devstack. Install using ``make install-local-ora``. 
2. Create an ORA block. On the settings tab check that ``Response Editor`` Dropdown is available.
3. Check that the ORA block working fine as before this implementation.

**Author notes and concerns**:

1. This implementation utilizes the availability of ``RequireJS`` on both LMS and Studio. 

**Reviewers**
